### PR TITLE
Refactor PoseGraph work queue optimization dispatching

### PR DESCRIPTION
**Pull request by danielsievers**
Unwrap the logic to dispatch optimization and the deferred
logic to create the work queue (when kicked off in foreground)
to happen in AddWorkItem() and HandleWorkQueue().
The work items will instead return whether they need optimization
to be dispatched or not.

### CODE REVIEW

File: cartographer/mapping/internal/3d/pose_graph_3d.h [hunk 1]
```diff
@@ -240,9 +241,6 @@ class PoseGraph3D : public PoseGraph {
   // Number of nodes added since last loop closure.
   int num_nodes_since_last_loop_closure_ GUARDED_BY(mutex_) = 0;
 
-  // Whether the optimization has to be run before more data is added.
```
[the remark this answers is not in the document]
Comment (danielsievers) [reply cartographer/mapping/internal/3d/pose_graph_3d.h]: oops, bad rebase. thanks. fixed.

**Comment by wohe** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user wohe. Merge queue now has a length of 1.

**Comment by wohe** [comment]
@gaschler Can you verify, please?

**Pull request merged**